datafusion/src/physical_optimizer/repartition.rs

```diff
@@ -36,33 +128,70 @@ impl Repartition {
     }
 }
 
+/// Recursively visits all `plan`s puts and then optionally adds a
+/// `RepartitionExec` at the output of `plan` to match
+/// `target_partitions`
+///
+/// if `can_reorder` is false, means that the output of this node
+/// can not be reordered as as something upstream is relying on that order
+///
+/// If 'would_benefit` is false, the upstream operator doesn't
+///  benefit from additional reordering
+///
 fn optimize_partitions(
     target_partitions: usize,
     plan: Arc<dyn ExecutionPlan>,
-    should_repartition: bool,
+    can_reorder: bool,
+    would_benefit: bool,
 ) -> Result<Arc<dyn ExecutionPlan>> {
-    // Recurse into children bottom-up (added nodes should be as deep as possible)
+    // Recurse into children bottom-up (attempt to repartition as
+    // early as possible)
 
     let new_plan = if plan.children().is_empty() {
         // leaf node - don't replace children
         plan.clone()
     } else {
-        let should_repartition_children = plan.should_repartition_children();
+        let can_reorder_children =
+            match (plan.relies_on_input_order(), plan.maintains_input_order()) {
+                (true, _) => {
+                    // `plan` itself relies on the order of its
+                    // children, so don't reorder them!
+                    false
+                }
+                (false, false) => {
+                    // `plan` may reorder the input itself, so no need
+                    // to preserve the order of any children
+                    true
```

Comment:
Interestingly when I add the `requres_single_partition` case here I get failures with the tests on

```
[
    "SortPreservingMergeExec: [c1@0 ASC]",
    "SortExec: [c1@0 ASC]",
    "ProjectionExec: expr=[c1@0 as c1]",
    "RepartitionExec: partitioning=RoundRobinBatch(10)",
    "ParquetExec: limit=None, partitions=[x]",
]
actual:

[
    "SortPreservingMergeExec: [c1@0 ASC]",
    "SortExec: [c1@0 ASC]",
    "ProjectionExec: expr=[c1@0 as c1]",
    "ParquetExec: limit=None, partitions=[x]",
]
```

Aka repartitioning doesn't two levels down.

So rather than intermix the "should we bother repartitioning" with the "would it produce wrong answers" I simply removed the check for the required input partitioning and it is now included in the default  "benefits from repartitioning check"